### ISSUE TITLE
UnsafeLegacyServerConnect option described in SSL_CONF_cmd man page has no effect?

### ISSUE THREAD

**Issue by mbali** [openssl/openssl#18295]
Affected version: 3.0.2

I don't really know if this is a documentation bug or a missing line in `ssl/ssl_conf.c`, but the `UnsafeLegacyServerConnect` option in the configuration does not seem to enable  `SSL_OP_LEGACY_SERVER_CONNECT`

`SSL_OP_LEGACY_SERVER_CONNECT` was set by default before OpenSSL 3.0.0, but now it is off.

I have to connect to an unpatched server with `curl`, and the connection failed with an SSL error:
`error:0A000152:SSL routines::unsafe legacy renegotiation disabled`

I tried to create a custom OpenSSL configuration file as a workaround, but setting the `UnsafeLegacyServerConnect` had no effect. As a temporary solution, I used the `UnsafeLegacyRenegotiation` option, which enabled me to connect to this server.

Will `UnsafeLegacyServerConnect` be supported in an upcoming version, or the man page will be updated? Or am I missing something else?

**Comment by t8m** [comment]
This is a bug, the option is documented but not implemented. I'd treat it as a bug rather than documentation issue.

**Comment by mikaello** [comment]
I couldn't find this bugfix in [CHANGES.md](https://github.com/openssl/openssl/blob/master/CHANGES.md), in which version of OpenSSL was this fixed?

**Comment by tom-cosgrove-arm** [comment]
TL;DR: 3.0.4.

As to how to find that out: from the `openssl-machine` comment above, it was fixed in commit `1e5d20b`

```
$ git tag --contains 1e5d20b        
openssl-3.0.4
openssl-3.0.5
openssl-3.0.6
openssl-3.0.7
openssl-3.0.8
openssl-3.1.0
```

So it was fixed in version 3.0.4.

If we want to confirm, the fix is

```
diff --git a/ssl/ssl_conf.c b/ssl/ssl_conf.c
index f49b2e5a37..ecd85e050c 100644
--- a/ssl/ssl_conf.c
+++ b/ssl/ssl_conf.c
@@ -384,6 +384,8 @@ static int cmd_Options(SSL_CONF_CTX *cctx, const char *value)
         SSL_FLAG_TBL_SRV("ECDHSingle", SSL_OP_SINGLE_ECDH_USE),
         SSL_FLAG_TBL("UnsafeLegacyRenegotiation",
                      SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION),
+        SSL_FLAG_TBL("UnsafeLegacyServerConnect",
+                     SSL_OP_LEGACY_SERVER_CONNECT),
         SSL_FLAG_TBL("ClientRenegotiation",
                      SSL_OP_ALLOW_CLIENT_RENEGOTIATION),
         SSL_FLAG_TBL_INV("EncryptThenMac", SSL_OP_NO_ENCRYPT_THEN_MAC),
```

and we can see that this change is indeed in the changes in `ssl/ssl_conf.c` from 3.0.3 to 3.0.4:

```
$ diff -u openssl-3.0.[34]/ssl/ssl_conf.c
--- openssl-3.0.3/ssl/ssl_conf.c	2022-05-03 14:32:01.000000000 +0100
+++ openssl-3.0.4/ssl/ssl_conf.c	2022-06-21 14:30:58.000000000 +0100
@@ -384,6 +384,8 @@
         SSL_FLAG_TBL_SRV("ECDHSingle", SSL_OP_SINGLE_ECDH_USE),
         SSL_FLAG_TBL("UnsafeLegacyRenegotiation",
                      SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION),
+        SSL_FLAG_TBL("UnsafeLegacyServerConnect",
+                     SSL_OP_LEGACY_SERVER_CONNECT),
         SSL_FLAG_TBL("ClientRenegotiation",
                      SSL_OP_ALLOW_CLIENT_RENEGOTIATION),
         SSL_FLAG_TBL_INV("EncryptThenMac", SSL_OP_NO_ENCRYPT_THEN_MAC),
@@ -700,7 +702,7 @@
     SSL_CONF_CMD_SWITCH("legacy_server_connect", SSL_CONF_FLAG_CLIENT),
     SSL_CONF_CMD_SWITCH("no_renegotiation", 0),
     SSL_CONF_CMD_SWITCH("no_resumption_on_reneg", SSL_CONF_FLAG_SERVER),
-    SSL_CONF_CMD_SWITCH("no_legacy_server_connect", SSL_CONF_FLAG_SERVER),
+    SSL_CONF_CMD_SWITCH("no_legacy_server_connect", SSL_CONF_FLAG_CLIENT),
     SSL_CONF_CMD_SWITCH("allow_no_dhe_kex", 0),
     SSL_CONF_CMD_SWITCH("prioritize_chacha", SSL_CONF_FLAG_SERVER),
     SSL_CONF_CMD_SWITCH("strict", 0),
```
